File: src/main/java/org/nitk/hackathon/bank/ControlPanel.java
package org.nitk.hackathon.bank;

public class ControlPanel {
    private enum State { CLOSED, OPEN, HALF_OPEN}

    private State state = State.CLOSED;

    private int failureCount = 0;
    private int failureThreshold = 3;
    private long openStateTimeout = 1000;
    private long lastFailureTime = 0;

    public synchronized boolean allowRequest() {
        if (state == State.OPEN) {
            if(System.currentTimeMillis() - lastFailureTime > openStateTimeout) {
                state = State.HALF_OPEN;
                return true;
            }
            return false;
        }
        return true;
    }

    public synchronized void recordSuccess() {
       state = State.CLOSED;
       failureCount = 0;
    }

    public synchronized void recordFailure() {
        failureCount++;
        if(failureCount >= failureThreshold) {
            state = State.OPEN;
            lastFailureTime = System.currentTimeMillis();
        }
    }
}

File: python/bank/control_panel.py

import time


class State:
    CLOSED, OPEN, HALF_OPEN = range(3)

class ControlPanel:
    def __init__(self):
        self.state = State.CLOSED
        self.failure_count = 0
        self.failure_threshold = 3
        self.open_state_timeout = 1000  # milliseconds
        self.last_failure_time = 0

    def allow_request(self):
        if self.state == State.OPEN:
            if (time.time() - self.last_failure_time) * 1000 > self.open_state_timeout:
                self.state = State.HALF_OPEN
                return True
            return False
        return True

    def record_success(self):
        self.state = State.CLOSED
        self.failure_count = 0

    def record_failure(self):
        self.failure_count += 1
        if self.failure_count >= self.failure_threshold:
            self.state = State.OPEN
            self.last_failure_time = time.time()
